src/mtype.d

```diff
@@ -7129,22 +7129,23 @@ public:
                  *      // TypeIdentifier 'a', 'e', and 'v' should be TOKvar,
                  *      // because getDsymbol() need to work in AliasDeclaration::semantic().
                  */
-                if (!v.type || !v.type.deco)
+                if (!v.type ||
+                    !v.type.deco && v.inuse)
                 {
                     if (v.inuse) // Bugzilla 9494
-                    {
-                        error(loc, "circular reference to '%s'", v.toPrettyChars());
-                        *pe = new ErrorExp();
-                        return;
-                    }
-                    if (v.sem < SemanticDone && v._scope)
-                        v.semantic(null);
+                        error(loc, "circular reference to %s '%s'", v.kind(), v.toPrettyChars());
+                    else
+                        error(loc, "forward reference to %s '%s'", v.kind(), v.toPrettyChars());
+                    *pt = Type.terror;
+                    return;
                 }
-                assert(v.type); // Bugzilla 14642
                 if (v.type.ty == Terror)
+                {
                     *pt = Type.terror;
-                else
-                    *pe = new VarExp(loc, v);
+                    return;
```

Comment:
No, it's not churm. Currently `Type.resolve` returns randomly returns `Type.terror` via `*pt` or `ErrorExp` via `*pe`. But essentially, all errors should be translated to `Type.terror` because the source object is `Type` class.